### PR TITLE
Fix Move operations and expand librapay to support user modules and scripts

### PR DESCRIPTION
**Pull request by jackcmay**
This issue corresponds with Solana issue https://github.com/solana-labs/solana/issues/6898

- Fixes the Move example and librapay helpers to support the new Libra CLI compiler behavior
- Adds support for publishing user modules
- Adds support for running user scripts

### CODE REVIEW

**Comment by jackcmay** [comment]
@jstarry Would love some feedback on my novice .js :-)

File: src/cli/main.js [hunk 1]
```diff
@@ -5,70 +5,244 @@
...
+  await loadScript(connection, scriptAccount, scriptPath);
+  console.log('Loaded script ' + scriptAccount.publicKey.toString());
+
+  // Run that script
+
+  console.log('Running script ' + scriptAccount.publicKey.toString());
+  const enlightenedAccount = await createAccount(connection, payerAccount);
+  const argsLayout = lo.struct([
+    lo.nu64('numArgs'),
+    lo.u32('addressType'),
```
Comment (jstarry) [inline src/cli/main.js]: `'argType'`?

File: src/program/instruction_data.js [hunk 2]
```diff
@@ -63,10 +72,45 @@ export function createGenesis(amount: number): Buffer {
...
+  ]);
+
+  var buffer = Buffer.alloc(layout.span);
+  layout.encode(
+    {
+      instruction: Instruction.InvokeMain,
+      length: 92, // TODO calculate
+      command: Command.RunScript,
+      senderAddress: senderPublicKey.toBuffer(),
+      functionNameLength: 4,
```
Comment (jstarry) [inline src/program/instruction_data.js]: Should this be `functionName.length`?

File: src/program/instruction_data.js [hunk 3]
```diff
@@ -63,10 +72,45 @@ export function createGenesis(amount: number): Buffer {
...
+      functionName: Buffer.from(functionName, 'utf8'),
```
Comment (jstarry) [inline src/program/instruction_data.js]: Should this be `'ascii'`?

File: src/program/instruction_data.js [hunk 4]
```diff
@@ -63,10 +72,45 @@ export function createGenesis(amount: number): Buffer {
...
+    publicKeyLayout('senderAddress'),
+    lo.nu64('functionNameLength'),
+    lo.blob(4, 'functionName'),
+  ]);
+
+  var buffer = Buffer.alloc(layout.span);
+  layout.encode(
+    {
+      instruction: Instruction.InvokeMain,
+      length: 92, // TODO calculate
```
Comment (jstarry) [inline src/program/instruction_data.js]: TODO still relevant?

File: src/program/librapay.js [hunk 5]
```diff
@@ -113,6 +89,105 @@ export async function createGenesis(
   return genesisAccount;
 }
 
+/**
+ * Load a new instance of a Move program on-chain.
+ * Returns the new program account.
```
Comment (jstarry) [inline src/program/librapay.js]: No longer relevant

File: src/program/librapay.js [hunk 6]
```diff
@@ -113,6 +89,105 @@ export async function createGenesis(
...
+ */
+export async function publishModule(
+  connection: Connection,
+  moduleAccount: Account,
+  path: string,
+): Promise<void> {
+  try {
+    await connection.getAccountInfo(moduleAccount.publicKey);
+  } catch (e) {
+    return await MoveLoader.load(
```
Comment (jstarry) [inline src/program/librapay.js]: If you remove the `await` here, you can avoid having the `await` throw an `UnhandledPromiseException`. This allows the caller of `publishModule` to properly handle an error from `MoveLoader.load(..)`

File: src/program/librapay.js [hunk 7]
```diff
@@ -113,6 +89,105 @@ export async function createGenesis(
...
+      moduleAccount,
+      path,
+    );
+  }
+  console.log('Warning: module already published or account already exists');
+}
+
+/**
+ * Load a new instance of a Move program on-chain.
+ * Returns the new program account.
```
Comment (jstarry) [inline src/program/librapay.js]: nit: looks like this method doesn't return a pubkey anymore

File: src/program/librapay.js [hunk 8]
```diff
@@ -113,6 +89,105 @@ export async function createGenesis(
...
+  return await MoveLoader.load(
+    connection,
+    AccountType.CompiledScript,
+    scriptAccount,
+    path,
+  );
+}
+
+/**
+ * Mint tokens into a new Libra account.
```
Comment (jstarry) [inline src/program/librapay.js]: nit: update comment

File: src/program/librapay.js [hunk 9]
```diff
@@ -113,6 +89,105 @@ export async function createGenesis(
...
+ * Returns the new account
+ */
+export async function runScript(
+  connection: Connection,
+  scriptAccount: PublicKey,
```
Comment (jstarry) [inline src/program/librapay.js]: nit: scriptPubkey

File: src/program/librapay.js [hunk 10]
```diff
@@ -113,6 +89,105 @@ export async function createGenesis(
...
+      senderAccount.publicKey,
+      functionName,
+      args,
+    ),
+  });
+
+  let signerAccounts = [payerAccount, senderAccount];
+  signerAccounts = signerAccounts.concat(additionalSignerAccounts);
+
+  await sendAndConfirmTransaction(
```
Comment (jstarry) [inline src/program/librapay.js]: As before, better to return the promise here

File: src/program/librapay.js [hunk 11]
```diff
@@ -113,6 +89,105 @@ export async function createGenesis(
...
+    programId: MoveLoader.programId,
+    data: InstructionData.runScript(
+      senderAccount.publicKey,
+      functionName,
+      args,
+    ),
+  });
+
+  let signerAccounts = [payerAccount, senderAccount];
+  signerAccounts = signerAccounts.concat(additionalSignerAccounts);
```
Comment (jstarry) [inline src/program/librapay.js]: In one line `let signerAccounts = [payerAccount, senderAccount, ...additionalSignerAccounts];`

File: src/program/librapay.js [hunk 12]
```diff
@@ -172,54 +249,56 @@ export async function pay(
   connection: Connection,
   payerAccount: Account,
   genesisAccount: Account,
-  senderAccount: Account,
+  senderAccountAccount: Account,
```
Comment (jstarry) [inline src/program/librapay.js]: should just be `senderAccount` right?

File: src/program/librapay.js [hunk 13]
```diff
@@ -172,54 +249,56 @@ export async function pay(
...
+        pubkey: scriptAccount.publicKey,
         isSigner: false,
-        isDebitable: false,
+        isWritable: false,
       },
       {
         pubkey: genesisAccount.publicKey,
         isSigner: false,
-        isDebitable: true,
+        isWritable: true,
```
Comment (jstarry) [inline src/program/librapay.js]: Why is genesis account writable here?

File: package.json [hunk 14]
```diff
@@ -31,7 +31,7 @@
   "dependencies": {
     "@material-ui/core": "^3.9.3",
     "@material-ui/icons": "^3.0.2",
-    "@solana/web3.js": "^0.22.0",
+    "@solana/web3.js": "^0.24.2",
```
Comment (jstarry) [inline package.json]: I think you need to be on `0.25.0` to get the `getAccountInfo` fix

File: programs/deps.json [hunk 15]
```diff
@@ -0,0 +1 @@
+[[76,73,66,82,65,86,77,10,1,0,7,1,74,0,0,0,2,0,0,0,3,76,0,0,0,3,0,0,0,13,79,0,0,0,5,0,0,0,14,84,0,0,0,2,0,0,0,5,86,0,0,0,18,0,0,0,4,104,0,0,0,32,0,0,0,11,136,0,0,0,17,0,0,0,0,0,0,1,0,2,1,2,0,0,3,0,1,77,15,117,110,105,118,101,114,115,97,108,95,116,114,117,116,104,27,47,73,9,110,62,93,189,15,207,169,192,192,205,146,217,171,59,33,84,75,52,213,221,74,101,217,139,135,139,153,34,0,1,0,1,0,2,0,6,42,0,0,0,0,0,0,0,2]]
```
[the remark this answers is not in the document]
Comment (jackcmay) [reply programs/deps.json]: Yes

File: src/program/librapay.js (hunk 11 above)
Comment (jackcmay) [reply src/program/librapay.js]: Cool

**Pull request merged**